### PR TITLE
fix(dsa): promote SM100 H16/H32 O*dO operands before multiplication

### PR DESCRIPTION
**Pull request by C-TC**
SM100 H16/H32 sparse-attention backward rounds `O*dO` in the input dtype before accumulating in FP32. Promote both operands before multiplying to avoid the extra rounding in the softmax delta and resulting gradients.

## Before submitting

- [x] I agree to license this contribution under the terms of LICENSE.txt.
- [x] I ran the repository pre-commit hooks and committed formatting changes.
- [x] I reviewed the applicable AGENTS.md Hard Rules.
- [x] Required labels (maintainer permission needed): `cat-bugfix`, `op: DSA`, `area:sparse_attention`, `orig-external`.

## Affected area

FE OSS kernels / CuTeDSL: SM100 sparse-attention backward.

## Summary

- In `dsa_bwd_sm100_h16.py`, convert O and dO fragments to the accumulator dtype before multiplication. H32 inherits this preprocessing.
- Extend the existing FP32-delta regression to H16/H32/H64 and check native specialization dispatch. Preserve its saved-BF16-output reference and numerical thresholds.

## Related issues

Completes operand promotion for the specializations alongside the generic-kernel fix in #877.

## API and compatibility impact

No interface, layout, dependency, or launch-geometry changes. The saved O remains BF16; the product and reduction use FP32.

## Testing

GB200, CUDA 13.2, PyTorch 2.11.0a0, CuTe DSL 4.6.2. Compared against develop `ebe3bba998b8`, changing only this kernel for the numerical control:

- Extended delta regression: unpatched **2 failed / 1 passed** (H16/H32 fail, H64 passes); patched **3 passed**.
- Specialization/invalid-slot/sink/partial-top-k regressions: **11 passed**, including the delta cases.
- Patched delta regression under Compute Sanitizer racecheck: **3 passed, zero hazards**.
- Changed-file Black/SPDX hooks passed.

Reproduce from `test/python`:

```bash
python -m pytest -o addopts='' -n 0 -v \
  fe_api/dsa/test_DSA_sparse_attention_backward.py \
  -k 'accumulates_odo or specialized_masks_invalid or specialized_handles_sink or partial_max_topk'
```

CUDA-graph microbenchmark, Q2048/KV4096/top-k2048/D576/V512, H16/32/64: matched A/B/A medians differ by less than 0.06%. Each arm uses warmup plus seven batches of 20 replays of the backward wrapper.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved numerical accuracy for sparse attention backward calculations by performing intermediate multiplication in FP32.

* **Tests**
  * Expanded regression coverage across configurations with 16, 32, and 64 attention heads.
  * Added validation that each configuration uses the expected native backend.

### CODE REVIEW

**Comment by vedaanta** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `625b65543ddf` against base `ebe3bba998b8`.

The operand promotion is correct, including H32 through its inherited preprocessing. No approval-blocking findings; retaining the existing nonblocking test-isolation comment below.

- **[P2] Still open:** [Tie the dispatch assertion to this test's invocation](https://github.com/NVIDIA/cudnn-frontend/pull/1134#discussion_r4048088692).

Validation: Independent SM100/CuTeDSL 4.7 run: 11 related specialization, sink, invalid-slot, partial-top-k and precision cases passed with zero skips. The three precision cases also passed alone in a fresh process. Against exact base ebe3bba998b8, the new H16/H32 cases failed and H64 passed (2 failed, 1 passed). Same-device base/head/base spot check at Q128/KV1024/top-k512/D576/V512, BF16, H16/H32/H64: checked all three gradients against the existing reference, isolated native dispatch, and CUDA Graph replay before timing. H16 GPU time was about 139.32 -> 139.86 us, H32 147.11 -> 147.70 us (roughly +0.4% each), and H64 183.41 -> 183.51 us. Warm CPU wrapper calls remained about 75-78 us with no material regression in this sample. Current-head pipeline 68627401 passed SM80, SM90 and both SM100 OSS shards. The remaining failures are the unchanged documentation-scan path and 11 Rubin gated-attention cuda-versus-cuda:0 declaration errors; those paths are outside this two-file diff. Style and current merge-metadata checks passed. No duplicate CI was requested.

Limitations: The independent GPU check was on one SM100 device, not GB200 or Rubin. I did not repeat the author's racecheck; performance numbers cover only the stated workload.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Review by Anerudhan — APPROVED** [review]
